Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,130 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+///
+/// An enumeration to describe the length of a piece of data in bytes.
+/// It is contained to lengths that can be converted to integer types.
+///
+public enum ByteLength {
+    case one
+    case two
+    case four
+    case eight
+}
+
+extension ByteLength {
+   
+    fileprivate var length: Int {
+
+        switch self {
+        case .one:
+            return 1
+        case .two:
+            return 2
+        case .four:
+            return 4
+        case .eight:
+            return 8
+        }
+    }
+}
+
+///
+/// A decoder that splits the received `ByteBuffer` by the number of bytes specified in a fixed length header
+/// contained within the buffer.
+/// For example, if you received the following four fragmented packets:
+///     +---+----+------+----+
+///     | A | BC | DEFG | HI |
+///     +---+----+------+----+
+///
+/// Given that the specified header length is 1 byte,
+/// where the first header specifies 3 bytes while the second header specifies 4 bytes,
+/// a `LengthFieldBasedFrameDecoder` will decode them into the following packets:
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private let lengthFieldLength: ByteLength
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else {
+            return .needMoreData
+        }
+
+        // convert the length field to an int specifying the length
+        guard let lengthFieldValue = self.frameLength(for: lengthFieldSlice) else {
+            return .needMoreData
+        }
+
+        guard let contentsFieldSlice = buffer.readSlice(length: lengthFieldValue) else {
```

Comment:
I don’t think this actually works. Let’s assume we get the following bytes (and length `.one`):

`5, 1, 2, 3`

Then nothing for a while, and then

`4, 5`

What we would expect is that the first `5` is the length (5 bytes) and the next 1, 2, 3, 4, 5 are the data.

With your code however, we’ll find the first 5 to be the length (correct), then we’ll try to read 5 bytes (correct but unavailable).
As soon as more bytes arrive, we’ll start the parsing loop again and we’ll interpret the `4` and another length field (incorrect).

I think you’ll want a inner type

```
enum State {
    case waitingForHeader
    case waitingForData(length: Int)
}
```

And `var state: State` and depending on the state you do the right thing in the parsing loop.

Reply:
Looks like a good point. Either way it needs a test. So, I will add that first and then fix the problem once confirmed.

Reply:
Code changes were needed. Good spot.